<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>EnrichmentAnalysis — EnrichmentAnalysis • RAmP</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="EnrichmentAnalysis — EnrichmentAnalysis"><meta property="og:description" content="Performs Gene Set Enrichment Analysis and Over Representation Analaysis using clusterProfiler package"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAmP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAmP.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/anuroopv/RAmP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>EnrichmentAnalysis</h1>
    <small class="dont-index">Source: <a href="https://github.com/anuroopv/RAmP/blob/HEAD/R/EnrichmentAnalysis.R" class="external-link"><code>R/EnrichmentAnalysis.R</code></a></small>
    <div class="hidden name"><code>EnrichmentAnalysis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs Gene Set Enrichment Analysis and Over Representation Analaysis using clusterProfiler package</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">EnrichmentAnalysis</span><span class="op">(</span></span>
<span>  enrich <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gsea"</span>, <span class="st">"ora"</span><span class="op">)</span>,</span>
<span>  nonExclusive.data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">Fraction</span>,</span>
<span>  rankBy <span class="op">=</span> <span class="st">"stat"</span>,</span>
<span>  KEGG <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>  padjustMethod <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  exclusive.data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  background <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  minGS <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  maxGS <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  simplify <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  simplify_cutoff <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  org <span class="op">=</span> <span class="st">"dme"</span>,</span>
<span>  <span class="va">contrasts</span>,</span>
<span>  pvalCutOff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  sigmaCutOff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  lfcCutOff <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>enrich</dt>
<dd><p>Can be either "gsea" or "enrich"</p></dd>


<dt>nonExclusive.data</dt>
<dd><p>Default is NULL. If not, the data should be given as list containing data frame(s)</p></dd>


<dt>Fraction</dt>
<dd><p>Can be either "Proteome" or "Enriched". Indicates the type of input data used</p></dd>


<dt>rankBy</dt>
<dd><p>Can be either stat or logFC. Genes are ranked accordingly for GSEA</p></dd>


<dt>KEGG</dt>
<dd><p>Logical, default is FALSE</p></dd>


<dt>ont</dt>
<dd><p>Can be "BP", "CC", "MF" or "ALL"</p></dd>


<dt>padjustMethod</dt>
<dd><p>Method for adjusting p-value (Refer clusterProfiler for available options)</p></dd>


<dt>exclusive.data</dt>
<dd><p>Default is NULL. If not, the data should be given as list containing data frame(s)</p></dd>


<dt>background</dt>
<dd><p>Logical, default is TRUE. If FALSE, background proteins will not be considered for over representation analysis</p></dd>


<dt>minGS</dt>
<dd><p>minimal size of annotated genes in geneset</p></dd>


<dt>maxGS</dt>
<dd><p>maximal size of annotated genes in geneset</p></dd>


<dt>simplify</dt>
<dd><p>Logical, default is FALSE. If TRUE, redundant GO terms will be simplified based on the simplify_cutoff</p></dd>


<dt>simplify_cutoff</dt>
<dd><p>Cutoff for simplify method. Numerical value from 0-1. Default is 0.7</p></dd>


<dt>org</dt>
<dd><p>Database of the organism. Drosophila melanogaster = "dme", Mus muscuslus ' "mmu", Homo sapiens = "hsa", Saccharomyces cerevisae = "sce". Default is "dme"</p></dd>


<dt>contrasts</dt>
<dd><p>Mentions the conditions to be compared. Ex. MUTANT_vs_WILDTYPE or (MUTANT-A_vs_WILDTYPE-A)_vs_(MUTANT-B_vs_WILDTYPE-B). This how the contrasts should be provided and the values should be the same as the one given in condition column of sampleTable. (MUTANT-A_vs_WILDTYPE-A)_vs_(MUTANT-B_vs_WILDTYPE-B): This type can be used for complex data when comparing the interaction between two conditions such genotype and time</p></dd>


<dt>pvalCutOff</dt>
<dd><p>P-value cut off for significant protein(s), site(s) and GO terms. Default is 0.05</p></dd>


<dt>sigmaCutOff</dt>
<dd><p>PI-value cut off for significant protein(s) and site(s). Default is 0.05 (Refer Xiao et al, 2014 and Hostrup et al, 2022 for details on PI-value)</p></dd>


<dt>lfcCutOff</dt>
<dd><p>Log-fold change cut off. Default is 0</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Generates list of significant GO terms</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Venkatasubramani Anuroop V, Imhof Axel, Lahiri Shibojyothi.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

